src/analysis/calculus/local_extr.lean

```diff
@@ -0,0 +1,276 @@
+
+lemma is_local_min.has_fderiv_at_eq_zero (h : is_local_min f a) (hf : has_fderiv_at f f' a) :
+  f' = 0 :=
+begin
+  suffices : ∀ v : E, (0:ℝ) ≤ f' v,
+  { ext v,
+    exact le_antisymm (by simpa using this (-v)) (this v) },
+  refine λ v, ge_of_tendsto at_top_ne_bot (hf.lim_real v) _,
+  apply mp_sets (mem_at_top (1:ℝ)),
+  have : tendsto (λ b:ℝ, a + b⁻¹ • v) at_top (𝓝 (a + (0:ℝ) • v)),
+    from tendsto_const_nhds.add (tendsto_smul tendsto_inverse_at_top_nhds_0 tendsto_const_nhds),
+  rw [zero_smul, add_zero] at this,
+  apply mem_sets_of_superset (mem_map.1 $ this h),
+  simp only [mem_set_of_eq, smul_eq_mul, mem_preimage],
+  assume c hfc hc,
+  exact mul_nonneg (le_trans zero_le_one hc) (sub_nonneg.2 hfc)
+end
+
+/-- The derivative at a local minimum equals zero. -/
+lemma is_local_min.fderiv_eq_zero (h : is_local_min f a) : fderiv ℝ f a = 0 :=
+if hf : differentiable_at ℝ f a then h.has_fderiv_at_eq_zero hf.has_fderiv_at
+else fderiv_zero_of_not_differentiable_at hf
+
+lemma is_local_max.has_fderiv_at_eq_zero (h : is_local_max f a) (hf : has_fderiv_at f f' a) :
+  f' = 0 :=
+neg_eq_zero.1 $ h.neg.has_fderiv_at_eq_zero hf.neg
+
+lemma is_local_max.fderiv_eq_zero (h : is_local_max f a) : fderiv ℝ f a = 0 :=
+if hf : differentiable_at ℝ f a then h.has_fderiv_at_eq_zero hf.has_fderiv_at
+else fderiv_zero_of_not_differentiable_at hf
+
+lemma is_local_extr.has_fderiv_at_eq_zero (h : is_local_extr f a) :
+  has_fderiv_at f f' a → f' = 0 :=
+h.elim is_local_min.has_fderiv_at_eq_zero is_local_max.has_fderiv_at_eq_zero
+
+lemma is_local_extr.fderiv_eq_zero (h : is_local_extr f a) : fderiv ℝ f a = 0 :=
+h.elim is_local_min.fderiv_eq_zero is_local_max.fderiv_eq_zero
+
+end vector_space
+
+section real
+
+variables {f : ℝ → ℝ} {f' : ℝ} {a b : ℝ}
+
+lemma is_local_min.has_deriv_at_eq_zero (h : is_local_min f a) (hf : has_deriv_at f f' a) :
+  f' = 0 :=
+by simpa using continuous_linear_map.ext_iff.1
+  (h.has_fderiv_at_eq_zero (has_deriv_at_iff_has_fderiv_at.1 hf)) 1
+
+lemma is_local_min.deriv_eq_zero (h : is_local_min f a) : deriv f a = 0 :=
+if hf : differentiable_at ℝ f a then h.has_deriv_at_eq_zero hf.has_deriv_at
+else deriv_zero_of_not_differentiable_at hf
+
+lemma is_local_max.has_deriv_at_eq_zero (h : is_local_max f a) (hf : has_deriv_at f f' a) :
+  f' = 0 :=
+neg_eq_zero.1 $ h.neg.has_deriv_at_eq_zero hf.neg
+
+lemma is_local_max.deriv_eq_zero (h : is_local_max f a) : deriv f a = 0 :=
+if hf : differentiable_at ℝ f a then h.has_deriv_at_eq_zero hf.has_deriv_at
+else deriv_zero_of_not_differentiable_at hf
+
+lemma is_local_extr.has_deriv_at_eq_zero (h : is_local_extr f a) :
+  has_deriv_at f f' a → f' = 0 :=
+h.elim is_local_min.has_deriv_at_eq_zero is_local_max.has_deriv_at_eq_zero
+
+lemma is_local_extr.deriv_eq_zero (h : is_local_extr f a) : deriv f a = 0 :=
+h.elim is_local_min.deriv_eq_zero is_local_max.deriv_eq_zero
+
+end real
+
+section MVT
+
+variables (f f' : ℝ → ℝ) {a b : ℝ} (hab : a < b) (hfc : continuous f) (hfI : f a = f b)
```

Comment:
You should only assume `continuous_on f (Icc a b)`.

Reply:
`exists_forall_le_of_compact_of_continuous` assumes `continuous f`.

Reply:
I mean, in order to prove this theorem with `continuous_on f (Icc a b)`, we need to prove `compact_image` with `continuous_on` assumption.

Reply:
Or, better, move this to the file`subset_properties.lean`, close to `compact_image` (just as you have done in your other PR for the image of connected sets).